hazelcast-client/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -189,86 +147,51 @@ public V join() {
         }
     }
 
-    private Object getResult() {
+    private Object resolveResponse(ClientMessage clientMessage, boolean deserialize) {
         if (defaultValue != null) {
             return defaultValue;
         }
 
-        // If value is already deserialized, use it.
-        if (deserializedValue != null) {
-            return deserializedValue;
-        }
-        if (deserializeResponse) {
-            // Otherwise, it is possible that received data may not be deserialized
-            // if "shouldDeserializeData" flag is not true in any of registered "DelegatingExecutionCallback".
-            // So, be sure that value is deserialized before returning to caller.
-            deserializedValue = serializationService.toObject(response);
-            return deserializedValue;
+        Object decodedResponse = decodeResponse(clientMessage);
+        if (deserialize) {
+            return serializationService.toObject(decodedResponse);
         }
-
-        return response;
-    }
-
-    private Object resolveMessageToValue(ClientMessage message) {
-        return clientMessageDecoder.decodeClientMessage(message);
+        return decodedResponse;
     }
 
-    protected void setError(Throwable error) {
-        this.error = error;
-    }
+    private Object decodeResponse(ClientMessage clientMessage) {
+        if (decodedResponse != VOID) {
+            return decodedResponse;
+        }
+        ClientMessage message = ClientMessage.createForDecode(clientMessage.buffer(), 0);
+        Object newDecodedResponse = clientMessageDecoder.decodeClientMessage(message);
```

Comment:
I have mixed feelings; so I'm fine with Mehmet's comment. It is up to the end user to not shoot himself in the foot and that is how it works with a regular future. I believe this is how it should be.

But would we be breaking any code that assumes each thread will get its own instance and now we return shared instance.